csharp/ql/src/semmle/code/csharp/Attribute.qll

```diff
@@ -51,9 +54,37 @@ class Attribute extends TopLevelExprParent, @attribute {
   /** Gets the element that this attribute is attached to. */
   Attributable getTarget() { attributes(this, _, result) }
 
-  /** Gets the `i`th argument of this attribute. */
+  /**
+   * Gets the `i`th argument of this attribute. This includes both constructor
+   * arguments and named arguments.
+   */
   Expr getArgument(int i) { result = this.getChildExpr(i) }
 
+  /**
+   * Gets the `i`th constructor argument of this attribute. For example, only
+   * `true` is a constructor argument in
+   *
+   * ```
+   * MyAttribute[true, Foo = 0]
+   * ```
+   */
+  Expr getConstructorArgument(int i) {
+    result = this.getArgument(i) and not exists(result.getExplicitArgumentName())
+  }
+
+  /**
+   * Gets the named argument `name` of this attribute. For example, only
+   * `0` is a named argument in
+   *
+   * ```
+   * MyAttribute[true, Foo = 0]
+   * ```
+   */
+  Expr getNamedArgument(string name) {
+    result = this.getArgument(_) and
+    result.getExplicitArgumentName() = name
```

Comment:
I think we should model this similar to expressions, where we use `getArgumentForParameter(Parameter)` or something, and make `getArgument()` also work in both cases where the argument is specified by name or position. This will make query writing easier as you won't need to check both cases every time.

Reply:
It is actually not the same in the case of attributes: Arguments to the constructor are always positional, and "named arguments" are in fact object initializers (see [this](https://github.com/Semmle/ql/pull/1171/files#diff-0a5e062897535fd7620d09825708fc02R68) example, where `x = 0` and `y = ""` are field/property initializers). This is why I wanted to distinguish the two cases (`getConstructorArgument()` vs `getNamedArgument()`). I did consider whether we should model `Attribute`s as sub classes of `ObjectCreation`, as that is essentially what it is, but it doesn't work nicely because there is no parent element.

Reply:
Ah ok, no it makes sense to treat the cases separately then.